course/page/upload.py

```diff
@@ -76,15 +91,151 @@ def clean_uploaded_file(self):
+
+    @staticmethod
+    def _get_uploaded_file_buf(files_data):
+        files_data["uploaded_file"].seek(0)
+        return files_data["uploaded_file"].read()
+
+    def files_data_to_answer_data(self, files_data):
+        buf = self._get_uploaded_file_buf(files_data)
+
+        if len(self.mime_types) == 1:
+            mime_type, = self.mime_types
+        else:
+            mime_type = files_data["uploaded_file"].content_type
+        from base64 import b64encode
+        return {
+                "base64_data": b64encode(buf).decode(),
+                "mime_type": mime_type,
+                }
+
+    def make_form(self, page_context, page_data,
+            answer_data, page_behavior):
+        form = FileUploadForm(
+                self.page_desc.maximum_megabytes, self.mime_types)
+        return form
+
+    def process_form_post(self, page_context, page_data, post_data, files_data,
+            page_behavior):
+        form = FileUploadForm(
+                self.page_desc.maximum_megabytes, self.mime_types,
+                self.form_clean_uploaded_file_callback,
+                post_data, files_data)
+        return form
+
+    def form_clean_uploaded_file_callback(self, uploaded_file):
+        """ This is used to handle uploaded_file field clean.
+        """
+        return uploaded_file
+
+    def get_form_to_html_context(self, request, page_context, form, answer_data):
+        ctx = {"form": form}
+        if answer_data is not None:
+            ctx["mime_type"] = answer_data["mime_type"]
+            ctx["data_url"] = "data:%s;base64,%s" % (
+                answer_data["mime_type"],
+                answer_data["base64_data"],
+                )
+            ctx["default_download_name"] = self.default_download_name
+        return ctx
+
+    def form_to_html(self, request, page_context, form, answer_data):
+        ctx = self.get_form_to_html_context(request, page_context, form, answer_data)
+        from django.template.loader import render_to_string
+        return render_to_string(self.form_template, ctx, request)
+
+    def answer_data(self, page_context, page_data, form, files_data):
+        return self.files_data_to_answer_data(files_data)
+
+    def get_download_file_extension(self):
+        if self.file_extension is not None:
+            return self.file_extension
+
+        ext = None
+        if len(self.mime_types) == 1:
+            mtype, = self.mime_types
+            from mimetypes import guess_extension
+            ext = guess_extension(mtype)
+
+        if ext is None:
+            ext = ".dat"
+        return ext
+
+    def normalized_bytes_answer(self, page_context, page_data, answer_data):
+        if answer_data is None:
+            return None
+
+        ext = self.get_download_file_extension()
+
+        from base64 import b64decode
+        return ext, b64decode(answer_data["base64_data"])
+
+
+class FileUploadQuestion(FileUploadQuestionBase):
```

Comment:
This is hard to review, as much of this code was moved. Could you outline the changes?

Reply:
I'll do that.